src/shim.h

```diff
@@ -18,44 +18,70 @@
  * This would be signal fd, stdin fd, proxy socket fd and an I/O
  * fd passed by the runtime
  */
-#define MAX_POLL_FDS 4
+#define MAX_POLL_FDS 3
 
 struct cc_shim {
 	char       *container_id;
 	int         proxy_sock_fd;
-	int         proxy_io_fd;
-	uint64_t    io_seq_no;
-	uint64_t    err_seq_no;
-	bool        exiting;
+	char       *token;
+	char       *proxy_address;
+	int         proxy_port;
 };
 
-/*
- * control message format
- * | ctrl id | length  | payload (length-8)      |
- * | . . . . | . . . . | . . . . . . . . . . . . |
- * 0         4         8                         length
- */
-#define CONTROL_HEADER_SIZE             8
-#define CONTROL_HEADER_LENGTH_OFFSET    4
-
-/*
- * stream message format
- * | stream sequence | length  | payload (length-12)     |
- * | . . . . . . . . | . . . . | . . . . . . . . . . . . |
- * 0                 8         12                        length
- */
-#define STREAM_HEADER_SIZE              12
-#define STREAM_HEADER_LENGTH_OFFSET     8
-
-#define PROXY_CTL_HEADER_SIZE           8
-#define PROXY_CTL_HEADER_LENGTH_OFFSET  0
-
-/*
- * Hyperstart is limited to sending this number of bytes to
- * a client.
- *
- * (This value can be determined by inspecting the hyperstart
- * source where hyper_event_ops->wbuf_size is set).
- */
-#define HYPERSTART_MAX_RECV_BYTES       10240
+// Header size is length of header in 32 bit words.
+#define  MIN_HEADER_SIZE    3
+
+// Minimum supported proxy version.
+#define  PROXY_VERSION        2
+
+// Sizes in bytes
+#define  VERSION_SIZE         2
+#define  HEADER_LEN_SIZE      1
 
+// Offsets expressed as byte offsets within the header
+#define  HEADER_LEN_OFFSET    2
+#define  RES_OFFSET           6
+#define  OPCODE_OFFSET        7
+#define  PAYLOAD_LEN_OFFSET   8
+#define  PAYLOAD_OFFSET       12
+
+struct frame_header {
+	uint16_t    version;
+	uint8_t     header_len;
+	uint8_t     err;
+	uint8_t     type;
+	uint8_t     opcode;
+	uint32_t    payload_len; 
+};
+
+struct frame {
+	struct   frame_header header;
+	uint8_t *payload;
+};
+
+enum frametype {
+	type_command = 0,
+	type_response,
+	type_stream,
+	type_notification
+};
+
+enum commandtype {
```

Comment:
I have used the values in calls to send_proxy_message() and checking the frame header elements for equality after reading a frame.